pages/en/lb4/Testing-Your-application.md

```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
+
+  describe('getDetails()', () => {
+    it('retrieves details of a product', async () => {
+      const controller = new ProductController(repository);
+      const findStub = repository.find as sinon.SinonStub;
+      findStub.resolves([{id: 1, name: 'Pen'}]);
+
+      const details = await controller.getDetails(1);
+
+      expect(details).to.containDeep({name: 'Pen'});
+      expect(findStub).to.be.calledWithMatch({where: {id: 1}});
+    });
+  });
+
+  function givenStubbedRepository() {
+    repository = sinon.createStubInstance(ProductRepository);
+  }
+});
+```
+
+### Unit-test your Models and Repositories
+
+In a typical LoopBack appliction, models and repositories are relying mostly on the behavior provided by the framework (`@loopback/repository` package) and there is no need to test the LoopBack's built-in functionality unit-test level (the functionality has been already verified by the tests in the framework). What does need new unit-tests is any additional application-specific API.
+
+For example, if our `Person` Model has properties `firstname`, `middlename` and `surname` and provides a function to obtain the full name, then we should write unit tests to verify the implementation of this additional method.
+
+```ts
+// test/unit/models/person.model.unit.ts
+import {Person} from '../../models/person.model.ts'
+import {expect} from '@loopback/testlab';
+
+describe('Person (unit)', () => {
+  // we recommend to group tests by method names
+  describe('getFullName()', () => {
+    it('uses all three parts when present', () => {
+      const person = new Person({
+        firstname: 'Jane',
+        middlename: 'Smith',
+        surname: 'Brown'
+      });
+
+      const fullName = person.getFullName();
+
+      expect(fullName).to.equal('Jane Smith Brown');
+    });
+
+    it('omits middlename when not present', () => {
+      // etc.
+    });
+  });
+});
+```
+
+Writing a unit test for a custom Repository methods is not straightforward because `CrudRepositoryImpl` is based on legacy loopback-datasource-juggler that was not designed with Dependency Injection in mind. We are recommending to use integration tests to verify the implementation of custom Repository methods, see [Test your Repositories against a real database](#test-your-repositories-against-a-real-database) in [Integration Testing](#integration-testing).
+
+### Unit-test your Sequence
+
+While it's possible to test a custom Sequence class in isolation, we think it's better to rely on acceptance-level tests in this exceptional case. The reason is that a custom Sequence class typically has many dependencies (which makes test setup too long and complex), and at the same time it provides very little functionality on top of the injected sequence actions. Bugs are much more likely to caused by the way how the real sequence action implementations interact together (which is not covered by unit tests), instead of the sequence code itself (which is the only thing covered).
```

Comment:
Break up into multiple lines.

Reply:
`Bugs are much more likely to caused by the way how the real sequence action implementations interact together (which is not covered by unit tests), instead of the sequence code itself (which is the only thing covered).`

-->
`Bugs are much more likely to exist between interacting actions within a sequence (which is not covered by unit tests), instead of the sequence code itself.`